### PR TITLE
ISSUE-123 Allow download of encrypted attachments

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

**Comment by Arsnael** [comment]
```
[ERROR] Errors: 
[ERROR]   OpenDistroListeningMessageSearchIndexTest$RetrieveIndexedFlags.retrieveIndexedFlagsShouldRetrieveSystemFlags » NoClassDefFound
[ERROR]   OpenDistroListeningMessageSearchIndexTest$RetrieveIndexedFlags.retrieveIndexedFlagsShouldReturnAllSystemFlagsWhenAllFlagsSet » NoClassDefFound
[ERROR]   OpenDistroListeningMessageSearchIndexTest$RetrieveIndexedFlags.retrieveIndexedFlagsShouldReturnEmptyFlagsWhenNoFlags » NoClassDefFound
[ERROR]   OpenDistroListeningMessageSearchIndexTest$RetrieveIndexedFlags.retrieveIndexedFlagsShouldReturnEmptyWhenNotFound » NoClassDefFound
[ERROR]   OpenDistroListeningMessageSearchIndexTest$RetrieveIndexedFlags.retrieveIndexedFlagsShouldReturnUserAndSystemFlags » NoClassDefFound
[ERROR]   OpenDistroListeningMessageSearchIndexTest$RetrieveIndexedFlags.retrieveIndexedFlagsShouldReturnUserFlags » NoClassDefFound
[ERROR]   OpenDistroListeningMessageSearchIndexTest.addShouldBeIndempotent » NoClassDefFound
[ERROR]   OpenDistroListeningMessageSearchIndexTest.addShouldIndexEmailBodyWhenNotIndexableAttachment » NoClassDefFound
[ERROR]   OpenDistroListeningMessageSearchIndexTest.addShouldIndexMessageWithAttachment » NoClassDefFound
[ERROR]   OpenDistroListeningMessageSearchIndexTest.addShouldIndexMessageWithoutAttachment » NoClassDefFound
[ERROR]   OpenDistroListeningMessageSearchIndexTest.addShouldIndexMultipleMessages » NoClassDefFound
[ERROR]   OpenDistroListeningMessageSearchIndexTest.addShouldPropagateExceptionWhenExceptionOccurs » ExceptionInInitializer
[ERROR]   OpenDistroListeningMessageSearchIndexTest.deleteAllShouldNotThrowWhenEmptyIndex » NoClassDefFound
[ERROR]   OpenDistroListeningMessageSearchIndexTest.deleteAllShouldRemoveAllIndexes » NoClassDefFound
[ERROR]   OpenDistroListeningMessageSearchIndexTest.deleteShouldBeIdempotent » NoClassDefFound
[ERROR]   OpenDistroListeningMessageSearchIndexTest.deleteShouldNotThrowOnUnknownMessageUid » NoClassDefFound
[ERROR]   OpenDistroListeningMessageSearchIndexTest.deleteShouldOnlyRemoveIndexesPassedAsArguments » NoClassDefFound
[ERROR]   OpenDistroListeningMessageSearchIndexTest.deleteShouldPropagateExceptionWhenExceptionOccurs » NoClassDefFound
[ERROR]   OpenDistroListeningMessageSearchIndexTest.deleteShouldRemoveIndex » NoClassDefFound
[ERROR]   OpenDistroListeningMessageSearchIndexTest.deleteShouldRemoveMultipleIndexes » NoClassDefFound
[ERROR]   OpenDistroListeningMessageSearchIndexTest.deserializeElasticSearchListeningMessageSearchIndexGroup » NoClassDefFound
[ERROR]   OpenDistroListeningMessageSearchIndexTest.updateShouldBeIdempotent » NoClassDefFound
[ERROR]   OpenDistroListeningMessageSearchIndexTest.updateShouldNotUpdateNorThrowOnUnknownMessageUid » NoClassDefFound
[ERROR]   OpenDistroListeningMessageSearchIndexTest.updateShouldPropagateExceptionWhenExceptionOccurs » NoClassDefFound
[ERROR]   OpenDistroListeningMessageSearchIndexTest.updateShouldUpdateIndex » NoClassDefFound
```

**Comment by chibenwa** [comment]
RFFFFFFF

**Comment by Arsnael** [comment]
Will try to fix it here: https://github.com/linagora/tmail-backend/pull/185

File: tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/EncryptedEmailContent.scala [hunk 1]
```diff
@@ -71,6 +73,33 @@ object AttachmentMetadata {
       size = parsedAttachment.getContent.size())
 }
 
+object EncryptedAttachmentBlobId {
```
Comment (vttranlina) [inline tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/EncryptedEmailContent.scala:76]: good job, I should do that in my pr

File: tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraEncryptedEmailFastViewGetMethodContract.scala [hunk 2]
```diff
@@ -468,6 +469,29 @@ trait LinagoraEncryptedEmailFastViewGetMethodContract {
       .isEqualTo(MESSAGE_PREVIEW)
   }
 
+  @Test
```
Comment (vttranlina) [inline tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraEncryptedEmailFastViewGetMethodContract.scala:472]: Should we write a separate contract class for this?
Then we can extends it in both 
`LinagoraEncryptedEmailFastViewGetMethodContract.scala`,
`LinagoraEncryptedEmailDetailedViewGetMethodContract.scala`

**Review by vttranlina — APPROVED** [review]

File: tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraEncryptedEmailFastViewGetMethodContract.scala (hunk 2 above)
Comment (chibenwa) [reply tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraEncryptedEmailFastViewGetMethodContract.scala:472]: Mmm IMO no.

Separate contract is IMO overkill, duplication is IMO not needed.

**Comment by vttranlina** [comment]
I have a question. 
We have an object storage system, Why we don't generate a signed-URL, that clients can download directly blob. (client - S3 system, without James middle). Then the downloading make fewer costs James's bandwidth.

**Review by Arsnael — APPROVED** [review]

**Comment by chibenwa** [comment]
> Why we don't generate a signed-URL, that clients can download directly blob. (client - S3 system, without James middle). Then the downloading make fewer costs James's bandwidth.

That wouldn't fit straight away into JMAP API model to be honnest :-/

But sure it could be studdied for extensions.

**Pull request merged**